packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -0,0 +1,510 @@
+/// [Map] of [children]. The type T is the type of the keys used
+/// to identify each widget and determine which widget is selected. As
+/// required by the [Map] class, keys must be of consistent types
+/// and must be comparable. The ordering of the keys will determine the order
+/// of the widgets in the segmented control.
+///
+/// When the state of the segmented control changes, the widget calls the
+/// [onValueChanged] callback. The map key associated with the newly selected
+/// widget is returned in the [onValueChanged] callback. Typically, widgets
+/// that use a segmented control will listen for the [onValueChanged] callback
+/// and rebuild the segmented control with a new [groupValue] to update which
+/// option is currently selected.
+///
+/// The [children] will be displayed in the order of the keys in the [Map].
+/// The height of the segmented control is determined by the height of the
+/// tallest widget provided as a value in the [Map] of [children].
+/// The width of the segmented control is determined by the horizontal
+/// constraints on its parent. The available horizontal space is divided by
+/// the number of provided [children] to determine the width of each widget.
+/// The selection area for each of the widgets in the [Map] of
+/// [children] will then be expanded to fill the calculated space, so each
+/// widget will appear to have the same dimensions.
+///
+/// See also:
+///
+///  * <https://developer.apple.com/design/human-interface-guidelines/ios/controls/segmented-controls/>
+class SegmentedControl<T> extends StatefulWidget {
+  /// Creates an iOS-style segmented control bar.
+  ///
+  /// The [children] and [onValueChanged] arguments must not be null. The
+  /// [children] argument must be an ordered [Map] such as a [LinkedHashMap].
+  /// Further, the length of the [children] list must be greater than one.
+  ///
+  /// Each widget value in the map of [children] must have an associated key
+  /// that uniquely identifies this widget. This key is what will be returned
+  /// in the [onValueChanged] callback when a new value from the [children] map
+  /// is selected.
+  ///
+  /// The [groupValue] must be one of the keys in the [children] map.
+  /// The [groupValue] is the currently selected value for the segmented control.
+  /// If no [groupValue] is provided, or the [groupValue] is null, no widget will
+  /// appear as selected.
+  SegmentedControl({
+    Key key,
+    @required this.children,
+    @required this.onValueChanged,
+    this.groupValue,
+  })  : assert(children != null),
+        assert(children.length >= 2),
+        assert(onValueChanged != null),
+        assert(groupValue == null || children.keys.any((T child) => child == groupValue)),
+        super(key: key);
+
+  /// The identifying keys and corresponding widget values in the
+  /// segmented control.
+  ///
+  /// The map must have more than one entry.
+  /// This attribute must be an ordered [Map] such as a [LinkedHashMap].
```

Comment:
This is unenforceable and pretty hard to debug. Just make this take a LinkedHashMap if the implementation details depends on it.

Reply:
Hi, I'll be making a new PR to address these comments! As far as this specific comment, Greg and Ian discussed this and determined we should leave it as a LinkedHashMap - does their reasoning make sense, or do you think we should still change this?:

@Hixie
Hixie on Jun 15  Member
this should be LinkedHashMap, then you can drop the assert.

@Hixie
Hixie on Jun 15  Member
oh never mind, const { } doesn't give you a LinkedHashMap. Just remove the assert above and leave this as a Map.

@gspencergoog
gspencergoog on Jun 15  Member
Why remove the assert? Since the implementation depends on it being an ordered Map, it seems like you want to make sure that it really is one, and if the underlying representation were to change, you want it to fail.

Or are you saying that this won't work because it'll assert with a const <T, Widget>{}? Is that still ordered (I assume so)?

@Hixie
Hixie on Jun 17  Member
const maps are some other internal class that doesn't implement LinkedHashMap. I'm pretty sure they're still ordered though.

Reply:
I don't think you need to do anything here. Hixie's last point means that you can't enforce it as a LinkedHashMap, because then people can't use const Maps.  As long as they're ordered, it'll work, and that is unlikely to change anytime soon. Do leave in the comment though, since that's still part of the (unenforced) contract.

Reply:
@Hixie https://github.com/dart-lang/sdk/blob/8df84c0a00fb7af3eae8a9fbed6bbd325076e3df/runtime/lib/map_patch.dart#L15 this is ok no?

Reply:
Nevermind, I'm told it's this instead https://github.com/dart-lang/sdk/blob/08f59e5de360f53e370f4899144fe58f3bce5275/runtime/lib/immutable_map.dart#L10. 

Unfortunately we can't type check.